lib/puppet/application/facts.rb

```diff
@@ -6,4 +6,17 @@ class Puppet::Application::Facts < Puppet::Application::IndirectionBase
   # request to a specific environment on a master. There is no way to set this
   # behavior per-action, so it must be set for the face as a whole.
   environment_mode :not_required
+
+  def parse_options
+    @type = Puppet::Util::ConstantInflector.constant2file(self.class.name.to_s.sub(/.+:/, '')).to_sym
+    @face = Puppet::Face[@type, :current]
+
+
+    if command_line.args.include?('ps')
+      @action = Puppet::Face.find_action(@face.name, :ps)
+      @face   = @action.face if @action
+    else
+      super
+    end
+  end
```

Comment:
Curious why this is needed? I'd expect the face system to automagically resolve the action like it does for `puppet facts upload`.

Reply:
In order to redirect all the options to facter, we don't want to let puppet parse them.
If we let puppet to parse the options we have to define them or else, puppet won't know about them:
`puppet facts show os --yaml`
will result in :
`Error: Could not parse application options: missing argument: --yaml`

Reply:
Ah ok I see. So I think we're going to have problems if we take this approach of trying to pass arbitrary options to facter this way.

One problem is that puppet applications allow settings to be specified on the command line. So we still need to support something like `puppet facts show --environment dev`. That way you can collect custom facts from modules in a local environment on disk.

We're also going to have issues with command line arguments that are both puppet settings and facter options like `color` and `strict`. For example, with the current PR `puppet facts show --color false` won't affect output emitted by puppet. Also puppet's `color` setting tristate (false, html, ansi/true), while it's a boolean in facter. Similarly puppet's `strict` setting is an enumeration, but it's boolean in facter.

Another issue is that `puppet facts --help` won't be able to print which options it accepts since only facter knows. That introduces a bit of a UX discovery issue, and forces people to run `facter --help`.

I think it would be better for `puppet facts` to define the set of application options that are supported by the current version of facter that puppet depends on, such as `--custom-dir`, `--timing`, etc. Then have the `facts` application do the normal `parse_options`, which will set puppet settings and application options. Then in the `show` method, configure facter based on the current set of puppet settings and application options, for example:

 ```ruby
argv = []
argv << '--custom-dir' << options[:customdir] if options[:customdir]
...
if options[:timing]
  argv << '--timing'
else
  argv << '--no-timing'
end
...
argv << '--no-color' unless Puppet[:color]
...
Facter.init_cli_options(argv)
Facter.search_external(options[:external_dir])
```

It does mean there will be some duplication between the `puppet facts` application and `facter` but I don't think that it's a bad thing. If facter adds a new command line option in the future, then we can update the `facts` application and bump the facter gem dependency if/when that time comes.